Wemo Dimmer Support

angel12 opened the pull request:
Add support for the Wemo Dimmer Switch

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4111


## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

angel12 committed:
Wemo Dimmer Support

Add support for the Wemo Dimmer Switch

angel12 committed:
Add newline at end of file

Re: findings from hound

angel12 committed:
Syntax for the hound

Sorry for the excess edits, new to python

MartinHjelmare commented on homeassistant/components/light/wemo.py:
```diff
@@ -140,3 +148,89 @@ def turn_off(self, **kwargs):
+        self._model_name = self.wemo.model_name
+
+        wemo = get_component('wemo')
+        wemo.SUBSCRIPTION_REGISTRY.register(self.wemo)
```
Component resources should preferably be stored in the dict `hass.data`. We should consider if a refactor is appropriate in this PR.

angel12 replied:
The majority of my code for adding the dimmer was actually written by others, just did some copy pasting from the switch/wemo.py to get it working. I'm definitely new to python, and coding in general, so I am not so sure I am the one you would want to try to get this cleaned up, because it would have to be cleaned up for the switch component as well, right?

MartinHjelmare replied:
Yeah, ok. Then follow the rest of my comments, but leave the global attribute access.

MartinHjelmare commented on homeassistant/components/light/wemo.py:
```diff
@@ -140,3 +148,89 @@ def turn_off(self, **kwargs):
+        """Return true if dimmer is on. Standby is on."""
+        return self._state
+
+    def update(self):
```
Do you still need to have polling active, since you register a callback for updates in pywemo? Maybe you should have the property `should_poll` return `False`?

angel12 replied:
Added this in my recent update to the code

MartinHjelmare replied:
Now you can remove this method, right?

MartinHjelmare commented on homeassistant/components/light/wemo.py:
```diff
@@ -140,3 +148,89 @@ def turn_off(self, **kwargs):
+
+        wemo = get_component('wemo')
+        wemo.SUBSCRIPTION_REGISTRY.register(self.wemo)
+        wemo.SUBSCRIPTION_REGISTRY.on(self.wemo, None, self._update_callback)
```
Move the registration of the callback to the coroutine `async_added_to_hass`, to avoid a race where the entity hasn't been added to home assistant before calling the callback.

MartinHjelmare commented on homeassistant/components/light/wemo.py:
```diff
@@ -140,3 +148,89 @@ def turn_off(self, **kwargs):
+        updated = self.wemo.subscription_update(_type, _params)
+        self._update(force_update=(not updated))
+
+        if not hasattr(self, 'hass'):
```
Remove this after moving the registration of the callback.

MartinHjelmare commented on homeassistant/components/light/wemo.py:
```diff
@@ -140,3 +148,89 @@ def turn_off(self, **kwargs):
+            brightness = int((self._brightness / 255) * 100)
+        else:
+            brightness = 255
+        self.wemo.set_brightness(brightness)
```
There's no direct update of home assistant state in the `turn_on` method. Is it not needed? Compare with the `turn_off` method.

MartinHjelmare replied:
When the `self.wemo.set_brightness(brightness)` call is made, isn't the callback from wemo called in turn? If so, then you should not update state here, but late the callback do that. There's already a call to `self.schedule_update_ha_state()` in the callback. So if all this is true, then you should remove the `self.schedule_update_ha_state()` from `turn_on` and from `turn_off`.

angel12 replied:
Removed these lines in my recent update as well

MartinHjelmare commented on homeassistant/components/wemo.py:
```diff
@@ -14,19 +14,20 @@
     'LightSwitch': 'switch',
-    'CoffeeMaker': 'switch'
+    'Dimmer': 'light',
+    'CoffeeMaker': 'switch',
```
Decide on one way to sort this. Either have the keys sorted 🔤 or group the components. Or don't change the sorting and just append to the end.

angel12 replied:
Cleaning this up right now on my dev machine, will publish the update after I have tested it

MartinHjelmare commented on homeassistant/components/light/wemo.py:
```diff
@@ -140,3 +148,89 @@ def turn_off(self, **kwargs):
+        self._brightness = None
+        self._state = None
+        # look up model name once as it incurs network traffic
+        self._model_name = self.wemo.model_name
```
If this uses I/O move it to the `async_added_to_hass` coroutine and schedule it on the executor.

MartinHjelmare reviewed (changes requested):
Make sure no entity properties perform I/O when accessed. See below for further comments.

angel12 committed:
Change order of Models

Fixed order back to ABC order

angel12 committed:
Changes as requested

I made the changes I was comfortable with at this point, the dimmer addition now very closely mirrors what is under the switch component, at least as far as the parts necessary for the dimmer.

Any changes past these with regards to the subscription registry / callback info is probably going to be over my head, but I will try to look deeper at them if required.

Blackbaud-JasonCusati commented:
I'd love to see this work, what else needs to be done to get this pull request completed?

MartinHjelmare commented:
My comments above still need to be addressed, except the refactor to use `hass.data`.

angel12 commented:
Sorry, haven't had time to make the rest of the corrections. I should have some time this weekend though

angel12 committed:
Remove unnecessary lines

Removed self.schedule_update_ha_state() from turn off / turn on

angel12 commented:
@MartinHjelmare I know that this PR is not where you would prefer, but I think I have gone as far as I can now. As far as changing the hass.() stuff, it's something I would be interested in learning / figuring out, but as I said in my earlier comment, I am probably not at that level yet. 

Please let me know if there are any other changes that I would need to make to get this PR merged. Thanks again!

angel12 committed:
Remove update(self)

Removed update method

MartinHjelmare committed:
Move subscription to async_added_to_hass

* Move subscription.
* Clean up.

MartinHjelmare commented:
@angel12 I've pushed a commit with the outstanding changes from my comments. Please test to make sure everything works.

angel12 commented:
@MartinHjelmare sorry for the delay, holidays and such.

Just loaded it up, and everything looks good on my end.

MartinHjelmare commented:
Great!

I think there's a small fix I'd like to add to make sure the registration of the device is finished before the next call to register the callback. We also need to fix the merge conflict. I'll try get that done later today.

MartinHjelmare committed:
Wait until the job in the executor is done

MartinHjelmare committed:
Run gen_requirements_all script

MartinHjelmare committed:
Only update instance attributes via callback

MartinHjelmare reviewed (approved):
Can be merged when tests pass.
